This is fine except for flake8
```
honeybadgermpc/ipc.py:150:5: F401 '.passive.generate_test_zeros' imported but unused
honeybadgermpc/ipc.py:150:5: F401 '.passive.generate_test_triples' imported but unused
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -4,7 +4,7 @@ ENV PYTHONUNBUFFERED=1
 
 RUN apk --update add make vim
 
-RUN apk --update add gcc musl-dev gmp-dev
+RUN apk --update add gcc musl-dev gmp-dev tmux
 
 RUN mkdir -p /usr/src/HoneyBadgerMPC
 WORKDIR /usr/src/HoneyBadgerMPC

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -8,6 +8,11 @@ services:
     volumes:
       - .:/usr/src/HoneyBadgerMPC
     command: pytest -v --cov=honeybadgermpc
+    extra_hosts:
+      - "hbmpc_0:127.0.0.1"
+      - "hbmpc_1:127.0.0.1"
+      - "hbmpc_2:127.0.0.1"
+      - "hbmpc_3:127.0.0.1"
   builddocs:
     build:
       context: .

--- a/honeybadgermpc/field.py
+++ b/honeybadgermpc/field.py
@@ -76,6 +76,10 @@ The reason for the slightly confusing error message is that ``x`` and
 from gmpy import mpz
 
 
+class FieldsNotIdentical(Exception):
+    pass
+
+
 class FieldElement(object):
     """Common base class for elements."""
 
@@ -90,9 +94,33 @@ class FieldElement(object):
     __long__ = __int__
 
 
+class GF(object):
+    _field_cache = {}
+
+    def __init__(self, modulus):
+        self.modulus = modulus
+
+    def __call__(self, value):
+        return GFElement(value, self)
+
+    @staticmethod
+    def get(modulus):
+        if modulus in GF._field_cache:
+            return GF._field_cache[modulus]
+
+        if not mpz(modulus).is_prime():
+            raise ValueError("%d is not a prime" % modulus)
+
+        gf = GF(modulus)
+        GF._field_cache[modulus] = gf
+        return gf
+
+
 class GFElement(FieldElement):
 
-    def __init__(self, value):
+    def __init__(self, value, gf):
+        self.modulus = gf.modulus
+        self.field = gf
         self.value = value % self.modulus
 
     def __add__(self, other):
@@ -103,10 +131,11 @@ class GFElement(FieldElement):
             # We can do a quick test using 'is' here since
             # there will only be one class representing this
             # field.
-            assert self.field is other.field, "Fields must be identical"
-            return GFElement(self.value + other.value)
+            if self.field is not other.field:
+                raise FieldsNotIdentical
+            return GFElement(self.value + other.value, self.field)
         except AttributeError:
-            return GFElement(self.value + other)
+            return GFElement(self.value + other, self.field)
 
     __radd__ = __add__
 
@@ -115,48 +144,51 @@ class GFElement(FieldElement):
         if not isinstance(other, (GFElement, int)):
             return NotImplemented
         try:
-            assert self.field is other.field, "Fields must be identical"
-            return GFElement(self.value - other.value)
+            if self.field is not other.field:
+                raise FieldsNotIdentical
+            return GFElement(self.value - other.value, self.field)
         except AttributeError:
-            return GFElement(self.value - other)
+            return GFElement(self.value - other, self.field)
 
     def __rsub__(self, other):
         """Subtraction (reflected argument version)."""
-        return GFElement(other - self.value)
+        return GFElement(other - self.value, self.field)
 
     def __xor__(self, other):
         """Xor for bitvalues."""
         if not isinstance(other, (GFElement, int)):
             return NotImplemented
         try:
-            assert self.field is other.field, "Fields must be identical"
-            return GFElement(self.value ^ other.value)
+            if self.field is not other.field:
+                raise FieldsNotIdentical
+            return GFElement(self.value ^ other.value, self.field)
         except AttributeError:
-            return GFElement(self.value ^ other)
+            return GFElement(self.value ^ other, self.field)
 
     def __rxor__(self, other):
         """Xor for bitvalues (reflected argument version)."""
-        return GFElement(other ^ self.value)
+        return GFElement(other ^ self.value, self.field)
 
     def __mul__(self, other):
         """Multiplication."""
         if not isinstance(other, (GFElement, int)):
             return NotImplemented
         try:
-            assert self.field is other.field, "Fields must be identical"
-            return GFElement(self.value * other.value)
+            if self.field is not other.field:
+                raise FieldsNotIdentical
+            return GFElement(self.value * other.value, self.field)
         except AttributeError:
-            return GFElement(self.value * other)
+            return GFElement(self.value * other, self.field)
 
     __rmul__ = __mul__
 
     def __pow__(self, exponent):
         """Exponentiation."""
-        return GFElement(pow(self.value, exponent, self.modulus))
+        return GFElement(pow(self.value, exponent, self.modulus), self.field)
 
     def __neg__(self):
         """Negation."""
-        return GFElement(-self.value)
+        return GFElement(-self.value, self.field)
 
     def __invert__(self):
         """Inversion.
@@ -181,22 +213,23 @@ class GFElement(FieldElement):
             return (lastx, lasty, a)
 
         inverse = extended_gcd(self.value, self.modulus)[0]
-        return GFElement(inverse)
+        return GFElement(inverse, self.field)
 
     def __div__(self, other):
         """Division."""
         try:
-            assert self.field is other.field, "Fields must be identical"
+            if self.field is not other.field:
+                raise FieldsNotIdentical
             return self * ~other
         except AttributeError:
-            return self * ~GFElement(other)
+            return self * ~GFElement(other, self.field)
 
     __truediv__ = __div__
     __floordiv__ = __div__
 
     def __rdiv__(self, other):
         """Division (reflected argument version)."""
-        return GFElement(other) / self
+        return GFElement(other, self.field) / self
 
     __rtruediv__ = __rdiv__
     __rfloordiv__ = __rdiv__
@@ -216,7 +249,7 @@ class GFElement(FieldElement):
         # (congruent to 3 mod 4), there will be no reminder in the
         # division below.
         root = pow(self.value, (self.modulus+1)//4, self.modulus)
-        return GFElement(root)
+        return GFElement(root, self.field)
 
     def bit(self, index):
         """Extract a bit (index is counted from zero)."""
@@ -250,7 +283,8 @@ class GFElement(FieldElement):
     def __eq__(self, other):
         """Equality test."""
         try:
-            assert self.field is other.field, "Fields must be identical"
+            if self.field is not other.field:
+                raise FieldsNotIdentical
             return self.value == other.value
         except AttributeError:
             return self.value == other
@@ -258,7 +292,8 @@ class GFElement(FieldElement):
     def __ne__(self, other):
         """Inequality test."""
         try:
-            assert self.field is other.field, "Fields must be identical"
+            if self.field is not other.field:
+                raise FieldsNotIdentical
             return self.value != other.value
         except AttributeError:
             return self.value != other
@@ -266,7 +301,8 @@ class GFElement(FieldElement):
     def __cmp__(self, other):
         """Comparison."""
         try:
-            assert self.field is other.field, "Fields must be identical"
+            if self.field is not other.field:
+                raise FieldsNotIdentical
             # TODO Replace with (a > b) - (a < b)
             # see https://docs.python.org/3/whatsnew/3.0.html#ordering-comparisons
             return cmp(self.value, other.value)     # noqa  XXX until above is done
@@ -295,24 +331,6 @@ class GFElement(FieldElement):
         False
         """
         return self.value != 0
-
-
-class GF(object):
-    _field_cache = {}
-
-    @staticmethod
-    def get(modulus):
-        if modulus in GF._field_cache:
-            return GF._field_cache[modulus]
-
-        if not mpz(modulus).is_prime():
-            raise ValueError("%d is not a prime" % modulus)
-
-        GFElement.modulus = modulus
-        GFElement.field = GFElement
-
-        GF._field_cache[modulus] = GFElement
-        return GFElement
 
 
 def FakeGF(modulus):

--- a/honeybadgermpc/field.py
+++ b/honeybadgermpc/field.py
@@ -103,6 +103,9 @@ class GF(object):
     def __call__(self, value):
         return GFElement(value, self)
 
+    def __reduce__(self):
+        return (GF.get, (self.modulus,))
+
     @staticmethod
     def get(modulus):
         if modulus in GF._field_cache:

--- a/honeybadgermpc/ipc.py
+++ b/honeybadgermpc/ipc.py
@@ -1,46 +1,51 @@
-import _pickle as pickle
+import pickle
 import asyncio
 import sys
-import os
 import struct
-import socket
+from .passive import PassiveMpc
 
 
-class Sender(object):
-    def __init__(self):
-        self.receivers = {}
-        self.loop = asyncio.get_event_loop()
+class Senders(object):
+    def __init__(self, queues, config):
+        self.queues = queues
+        self.config = config
 
-    async def send(self, msg, ip, port):
-        data = pickle.dumps(msg)
-        padded_msg = struct.pack('>I', len(data)) + data
-        key = "%s:%d" % (ip, port)
+    async def connect(self):
+        # Setup all connections first before sending messages
+        streams = [asyncio.open_connection(
+                self.config[i].ip,
+                self.config[i].port
+            ) for i in range(len(self.queues))]
+        streams = await asyncio.gather(*streams)
+        writers = [stream[1] for stream in streams]
+
+        # Setup tasks to consume messages from queues
+        # This is to ensure that messages are delivered in the correct order
+        for i, q in enumerate(self.queues):
+            recvid = "%s:%d" % (config[i].ip, config[i].port)
+            asyncio.ensure_future(self.process_queue(writers[i], q, recvid))
+
+    async def process_queue(self, writer, q, recvid):
         try:
-            if key not in self.receivers:
-                receiver = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-                receiver.connect((ip, port))
-                receiver.setblocking(False)
-                self.receivers[key] = receiver
-            else:
-                receiver = self.receivers[key]
-
-            if receiver is not None:
-                await self.loop.sock_sendall(receiver, padded_msg)
-                # print('SEND %8s [%2d -> %s]' % (msg[1][0], msg[0], ip))
+            while True:
+                msg = await q.get()
+                if msg is None:
+                    break
+                # print('SEND %8s [%2d -> %s]' % (msg[1][0], msg[0], recvid))
+                data = pickle.dumps(msg)
+                padded_msg = struct.pack('>I', len(data)) + data
+                writer.write(padded_msg)
+                await writer.drain()
         except ConnectionResetError:
-            print("WARNING: Connection with peer [%s] reset." % ip)
-            self.receivers[key] = None
+            print("WARNING: Connection with peer [%s] reset." % recvid)
         except ConnectionRefusedError:
-            print("WARNING: Connection with peer [%s] refused." % ip)
-            self.receivers[key] = None
+            print("WARNING: Connection with peer [%s] refused." % recvid)
         except BrokenPipeError:
-            print("WARNING: Connection with peer [%s] broken." % ip)
-            self.receivers[key] = None
+            print("WARNING: Connection with peer [%s] broken." % recvid)
 
     def close(self):
-        for receiver in self.receivers.values():
-            if receiver is not None:
-                receiver.close()
+        for q in self.queues:
+            q.put_nowait(None)
 
 
 def handle_result(future):
@@ -53,41 +58,32 @@ def handle_result(future):
 
 class Listener(object):
     def __init__(self, q, port):
-        server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        server.bind(('', port))
-        server.listen(10)
-        server.setblocking(False)
-        self.server = server
         self.q = q
-        self.loop = asyncio.get_event_loop()
-        self.tasks = [self.loop.create_task(self.runServer())]
+        serverFuture = asyncio.start_server(self.handle_client, "", port)
+        self.serverTask = asyncio.ensure_future(serverFuture)
+        self.tasks = []
 
-    async def runServer(self):
+    async def handle_client(self, reader, writer):
+        self.tasks.append(asyncio.current_task())
+        # print("Received new connection", writer.get_extra_info("peername"))
         while True:
-            client, address = await self.loop.sock_accept(self.server)
-            # print("Received connection from", address)
-            self.tasks.append(self.loop.create_task(self.handle_client(client)))
-
-    async def handle_client(self, client):
-        while True:
-            raw_msglen = await self.recvall(client, 4)
+            raw_msglen = await self.recvall(reader, 4)
             if raw_msglen is None:
                 break
             msglen = struct.unpack('>I', raw_msglen)[0]
-            received_raw_msg = await self.recvall(client, msglen)
+            received_raw_msg = await self.recvall(reader, msglen)
             if received_raw_msg is None:
                 break
             received_msg = pickle.loads(received_raw_msg)
             # print('RECV %8s [from %2d]' % (received_msg[1][0], received_msg[0]))
             await self.q.put(received_msg)
 
-    async def recvall(self, sock, n):
+    async def recvall(self, reader, n):
         # Helper function to recv n bytes or return None if EOF is hit
         data = b''
         while len(data) < n:
-            packet = await self.loop.sock_recv(sock, n - len(data))
-            if not packet:
+            packet = await reader.read(n)
+            if len(packet) == 0:
                 return None
             data += packet
         return data
@@ -95,7 +91,10 @@ class Listener(object):
     async def getMessage(self):
         return await self.q.get()
 
-    def close(self):
+    async def close(self):
+        server = await self.serverTask
+        server.close()
+        await server.wait_closed()
         for task in self.tasks:
             task.cancel()
 
@@ -106,10 +105,11 @@ class NodeDetails(object):
         self.port = port
 
 
-def process_router(config, id):
-    sender = Sender()
-    q = asyncio.Queue()
-    listener = Listener(q, config[id].port)
+def setup_sockets(config, n, id):
+    senderQueues = [asyncio.Queue() for _ in range(n)]
+    sender = Senders(senderQueues, config)
+    listenerQueue = asyncio.Queue()
+    listener = Listener(listenerQueue, config[id].port)
 
     def makeSend(i):
         def _send(j, o):
@@ -117,10 +117,9 @@ def process_router(config, id):
             if i == j:
                 # If attempting to send the message to yourself
                 # then skip the network stack.
-                q.put_nowait((i, o))
+                listenerQueue.put_nowait((i, o))
             else:
-                loop = asyncio.get_event_loop()
-                loop.create_task(sender.send((i, o), config[j].ip, config[j].port))
+                senderQueues[j].put_nowait((i, o))
 
         return _send
 
@@ -134,25 +133,33 @@ def process_router(config, id):
     return (makeSend(id), makeRecv(id), sender, listener)
 
 
+async def runProgramAsProcesses(program, config, t, id):
+    send, recv, sender, listener = setup_sockets(config, len(config), id)
+    # Need to give time to the listener coroutine to start
+    #  or else the sender will get a connection refused.
+    await asyncio.sleep(1)
+    await sender.connect()
+    await asyncio.sleep(1)
+    context = PassiveMpc('sid', N, t, id, send, recv, program)
+    results = await asyncio.ensure_future(context._run())
+    await asyncio.sleep(1)
+    sender.close()
+    await listener.close()
+    await asyncio.sleep(1)
+    return results
+
+
 if __name__ == "__main__":
-    from .passive import runProgramAsProcesses
     from .passive import generate_test_zeros, generate_test_triples
     from .passive import test_prog1, test_prog2
 
-    # The "ALIAS" environment variable should be set.
-    # It should be of the form <prefix>_<party_id> for each of the parties.
     # N - total number of parties
     # t - total number of corrupt parties
     # port - port to be used for communication between parties
+    # host_id - Of the form <prefix>_<party_id>
     N, t, port = int(sys.argv[1]), int(sys.argv[2]), 7000
-    host = os.environ.get('ALIAS').split("_")
+    host = sys.argv[3].split("_")
     prefix, id = host[0] + "_", int(host[1])
-
-    # Use this config if you want to test it on multiple processes on the same server
-    # id = int(sys.argv[2])
-    # config = {
-    #     i: NodeDetails("localhost", p) for i, p in enumerate(range(port, port+n))}
-
     # Only one party needs to generate the initial shares
     if id == 0:
         print('Generating random shares of zero in sharedata/')
@@ -164,9 +171,8 @@ if __name__ == "__main__":
     loop = asyncio.get_event_loop()
     loop.set_debug(True)
     try:
-        config = {i: NodeDetails(prefix+str(i), port) for i in range(N)}
-        loop.run_until_complete(runProgramAsProcesses(test_prog1, config, N, t, id))
-        config = {i: NodeDetails(prefix+str(i), port-1000) for i in range(N)}
-        loop.run_until_complete(runProgramAsProcesses(test_prog2, config, N, t, id))
+        config = {i: NodeDetails(prefix+str(i), port+i) for i in range(N)}
+        loop.run_until_complete(runProgramAsProcesses(test_prog1, config, t, id))
+        loop.run_until_complete(runProgramAsProcesses(test_prog2, config, t, id))
     finally:
         loop.close()

--- a/honeybadgermpc/passive.py
+++ b/honeybadgermpc/passive.py
@@ -1,6 +1,6 @@
 import asyncio
 from asyncio import Future
-from .field import GF
+from .field import GF, GFElement
 from .polynomial import polynomialsOver
 from .router import simple_router
 from .ipc import process_router
@@ -148,11 +148,11 @@ def write_shares(f, modulus, degree, myid, shares):
 def shareInContext(context):
 
     def _binopField(fut, other, op):
-        assert type(other) in [ShareFuture, FieldFuture, Share, Field]
+        assert type(other) in [ShareFuture, GFElementFuture, Share, GFElement]
         if isinstance(other, ShareFuture) or isinstance(other, Share):
             res = ShareFuture()
-        elif isinstance(other, Field) or isinstance(other, FieldFuture):
-            res = FieldFuture()
+        elif isinstance(other, GFElement) or isinstance(other, GFElementFuture):
+            res = GFElementFuture()
 
         if isinstance(other, Future):
             def cb(_): return res.set_result(op(fut.result(), other.result()))
@@ -162,7 +162,7 @@ def shareInContext(context):
             fut.add_done_callback(cb)
         return res
 
-    class FieldFuture(Future):
+    class GFElementFuture(Future):
         def __add__(self, other): return _binopField(
             self, other, lambda a, b: a + b)
 
@@ -177,12 +177,12 @@ def shareInContext(context):
             # v is the local value of the share
             if type(v) is int:
                 v = Field(v)
-            assert type(v) is Field
+            assert type(v) is GFElement
             self.v = v
 
         # Publicly reconstruct a shared value
         def open(self):
-            res = FieldFuture()
+            res = GFElementFuture()
 
             def cb(f): return res.set_result(f.result())
             context.open_share(self).add_done_callback(cb)
@@ -192,7 +192,7 @@ def shareInContext(context):
         # TODO: add type checks for the operators
         # @typecheck(Share)
         def __add__(self, other):
-            if isinstance(other, Field):
+            if isinstance(other, GFElement):
                 return Share(self.v + other)
             elif isinstance(other, Share):
                 return Share(self.v + other.v)
@@ -212,7 +212,7 @@ def shareInContext(context):
         def __str__(self): return '{%d}' % (self.v)
 
     def _binopShare(fut, other, op):
-        assert type(other) in [ShareFuture, FieldFuture, Share, Field]
+        assert type(other) in [ShareFuture, GFElementFuture, Share, GFElement]
         res = ShareFuture()
         if isinstance(other, Future):
             def cb(_): return res.set_result(op(fut.result(), other.result()))
@@ -232,8 +232,8 @@ def shareInContext(context):
         def __mul__(self, other): return _binopShare(
             self, other, lambda a, b: a * b)
 
-        def open(self) -> FieldFuture:
-            res = FieldFuture()
+        def open(self) -> GFElementFuture:
+            res = GFElementFuture()
 
             def cb2(sh): return res.set_result(sh.result())
 

--- a/honeybadgermpc/passive.py
+++ b/honeybadgermpc/passive.py
@@ -3,7 +3,6 @@ from asyncio import Future
 from .field import GF, GFElement
 from .polynomial import polynomialsOver
 from .router import simple_router
-from .ipc import process_router
 import random
 
 
@@ -258,19 +257,6 @@ async def runProgramAsTasks(program, N, t):
         tasks.append(loop.create_task(context._run()))
 
     results = await asyncio.gather(*tasks)
-    return results
-
-
-async def runProgramAsProcesses(program, config, N, t, id):
-    loop = asyncio.get_event_loop()
-    send, recv, sender, listener = process_router(config, id)
-    # Need to give time to the listener coroutine to start
-    #  or else the sender will get a connection refused.
-    await asyncio.sleep(1)
-    context = PassiveMpc('sid', N, t, id, send, recv, program)
-    results = await loop.create_task(context._run())
-    sender.close()
-    listener.close()
     return results
 
 

--- a/honeybadgermpc/polynomial.py
+++ b/honeybadgermpc/polynomial.py
@@ -1,6 +1,7 @@
 import operator
 import random
 from functools import reduce
+from .field import GF, GFElement
 
 
 def strip_trailing_zeros(a):
@@ -43,7 +44,7 @@ def polynomialsOver(field):
             # shares are in the form (x, y=f(x))
             if type(x_recomb) is int:
                 x_recomb = field(x_recomb)
-            assert type(x_recomb) is field
+            assert type(x_recomb) is GFElement
             xs, ys = zip(*shares)
             vector = []
             for i, x_i in enumerate(xs):
@@ -60,7 +61,7 @@ def polynomialsOver(field):
             """
             n = len(ys)
             assert n & (n-1) == 0, "n must be power of two"
-            assert type(omega) is field
+            assert type(omega) is GFElement
             assert omega ** n == 1, "must be an n'th root of unity"
             assert omega ** (n//2) != 1, "must be a primitive n'th root of unity"
             coeffs = [b/n for b in fft_helper(ys, 1/omega, field)]
@@ -68,7 +69,7 @@ def polynomialsOver(field):
 
         def evaluate_fft(self, omega, n):
             assert n & (n-1) == 0, "n must be power of two"
-            assert type(omega) is field
+            assert type(omega) is GFElement
             assert omega ** n == 1, "must be an n'th root of unity"
             assert omega ** (n//2) != 1, "must be a primitive n'th root of unity"
             return fft(self, omega, n)
@@ -158,7 +159,6 @@ def fft(poly, omega, n, seed=None):
 
 
 if __name__ == "__main__":
-    from .field import GF
     field = GF.get(0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001)
     Poly = polynomialsOver(field)
     poly = Poly.random(degree=7)

--- a/honeybadgermpc/secretshare_functionality.py
+++ b/honeybadgermpc/secretshare_functionality.py
@@ -1,6 +1,6 @@
 import asyncio
 import random
-from .field import GF
+from .field import GF, GFElement
 from .polynomial import polynomialsOver
 
 """
@@ -37,7 +37,7 @@ class SecretShare_Functionality(object):
     async def _run(self):
         v = await self.inputFromDealer
         # TODO: allow v to be arbitrary strings, or a parameter?
-        assert type(v) is Field
+        assert type(v) is GFElement
         poly = Poly.random(self.f, y0=v)
         for i in range(self.N):
             # TODO: this needs to be made into an "eventually send"

--- a/launch_ipc.sh
+++ b/launch_ipc.sh
@@ -1,0 +1,16 @@
+#!/bin/bash
+# Assume hosts are as follows:
+"""
+ 127.0.0.1   hbmpc_0
+ 127.0.0.1   hbmpc_1
+ 127.0.0.1   hbmpc_2
+ 127.0.0.1   hbmpc_3
+"""
+
+CMD="python -m honeybadgermpc.ipc 4 1"
+set -x
+tmux new-session     "${CMD} hbmpc_0; sh" \; \
+     splitw -h -p 50 "${CMD} hbmpc_1; sh" \; \
+     splitw -v -p 50 "${CMD} hbmpc_2; sh" \; \
+     selectp -t 0 \; \
+     splitw -v -p 50 "${CMD} hbmpc_3; sh"

--- a/tests/test_field.py
+++ b/tests/test_field.py
@@ -1,0 +1,27 @@
+import pytest
+import operator
+
+
+def test_multiple_fields():
+    from honeybadgermpc.field import GF
+    Field1 = GF.get(17)
+    Field2 = GF.get(7)
+    assert Field1.modulus == 17
+    assert Field2.modulus == 7
+
+
+def test_invalid_operations_on_fields():
+    from honeybadgermpc.field import GF, FieldsNotIdentical
+    Field1, Field2 = GF.get(17), GF.get(7)
+    operators = [
+        operator.add,
+        operator.sub,
+        operator.xor,
+        operator.mul,
+        operator.truediv,
+        operator.floordiv,
+        operator.eq,
+    ]
+    for op in operators:
+        with pytest.raises(FieldsNotIdentical):
+            op(Field1(2), Field2(3))

